nxcomp/Log.h

```diff
@@ -0,0 +1,461 @@
+	} per_thread_data;
+
+
+	static void free_thread_data(void* arg)
+	{
+		per_thread_data *pdt = (per_thread_data*)arg;
+		delete pdt->current_file;
+		delete pdt->thread_name;
+		delete pdt->buffer;
+		free(pdt);
+	}
+
+	per_thread_data* get_data_int() const
+	{
+		per_thread_data *ret = NULL;
+ 
+		if ( (ret = (per_thread_data*)pthread_getspecific(tls_key_)) == NULL )
+		{
+			ret = (per_thread_data*)malloc(sizeof(*ret));
+			ret->current_level = NXDEBUG;
+			ret->buffer = new std::stringstream();
+			ret->current_file = new std::string();
+			ret->thread_name = new std::string();
+			pthread_setspecific(tls_key_, ret);
+		}
+
+		return ret;
+	}
+
+	per_thread_data* get_data()
+	{
+		return get_data_int();
+	}
+
+	const per_thread_data* get_data() const
+	{
+		return get_data_int();
+	}
+
+	/** Convert NXLogStamp to string according to the current configuration */
+	std::string stamp_to_string(const NXLogStamp& stamp) const;
+
+	public:
+	NXLog()
+	{
+		stream_             = &std::cerr;
+		level_              = NXDEBUG;
+		synchronized_       = true;
+		thread_buffer_size_ = 1024;
+		log_level_          = true;
+		log_time_           = true;
+		log_unix_time_      = false;
+		log_location_       = true;
+		log_thread_id_      = true;
+
+		pthread_key_create(&tls_key_, free_thread_data);
+
+	}
+
+	/** Minimum severity level to output */
+	NXLogLevel level() const
+	{
+		return level_;
+	}
+
+	void level(NXLogLevel level)
+	{
+		level_ = level;
+	}
+
+
+	/** Current severity level */
+	NXLogLevel current_level() const
+	{
+		return get_data()->current_level;
+	}
+
+	void current_level(NXLogLevel level)
+	{
+		get_data()->current_level = level;
+	}
+
+	/** Source file from which messages are currently originating */
+	std::string current_file() const
+	{
+		return *get_data()->current_file;
+	}
+
+	void current_file(std::string val)
+	{
+		*get_data()->current_file = val;
+	}
+
+	std::ostream* stream() const
+	{
+		return stream_;
+	}
+
+	void stream(std::ostream *stream)
+	{
+		flush();
+		stream_ = stream;
+	}
+
+	bool synchronized() const {
+		return synchronized_;
+	}
+
+	void synchronized(bool val) {
+		synchronized_ = val;
+	}
+
+	bool log_level() const
+	{
+		return log_level_;
+	}
+
+	void log_level(bool val)
+	{
+		log_level_ = val;
+	}
+
+	bool log_time() const 
+	{
+		return log_time_;
+	}
+
+	void log_time(bool val)
+	{
+		log_time_ = val;
+	}
+
+	bool log_unix_time() const
+	{
+		return log_unix_time_;
+	}
+
+	void log_unix_time(bool val)
+	{
+		log_unix_time_ = val;
+	}
+
+	bool log_location() const
+	{
+		return log_location_;
+	}
+
+	void log_location(bool val)
+	{
+		log_location_ = val;
+	}
+
+	bool log_thread_id() const
+	{
+		return log_thread_id_;
+	}
+
+	void log_thread_id(bool val)
+	{
+		log_thread_id_ = val;
+	}
+
+
+	void flush()
+	{
+		per_thread_data *pdt = get_data();
+		const std::string str = pdt->buffer->str();
+
+		if (!str.empty())
+		{
+			pthread_mutex_lock(&output_lock_);
+			(*stream()) << str;
```

Comment:
Then we'd have to detect and get rid of duplicates, or to edit every log action in the code to remove the \n from it, or everything is going to come out double spaced.